core/src/main/java/io/grpc/internal/SharingExecutorService.java

```diff
@@ -0,0 +1,119 @@
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Wraps an existing executor, that keeps track of only the tasks it has added.
+ */
+public final class SharingExecutorService extends AbstractExecutorService {
+  private final Map<CleanupTask, Future<?>> activeTasks = new HashMap<CleanupTask, Future<?>>();
+
+  private final ExecutorService delegate;
+  private boolean shutdown;
+
+  public SharingExecutorService(ExecutorService delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public synchronized void shutdown() {
+    shutdown = true;
+    if (activeTasks.isEmpty()) {
+      notifyAll();
+    }
+  }
+
+  @Override
+  public synchronized List<Runnable> shutdownNow() {
+    shutdown = true;
+    List<Runnable> neverRan = new ArrayList<Runnable>(activeTasks.size());
+    Iterator<Entry<CleanupTask, Future<?>>> entries = activeTasks.entrySet().iterator();
+    while (entries.hasNext()) {
+      Entry<CleanupTask, Future<?>> entry = entries.next();
+      // isDone() returns true before the done() method on CleanupTask is called, so we can still
+      // remove it.
+      if (entry.getValue().isDone()) {
+        entries.remove();
+        continue;
+      }
+      // Since the underlying executor may have started running this, this is best effort.
+      boolean cancelled = entry.getValue().cancel(true);
+      if (cancelled) {
+        // This will not be the same runnable submitted by the user.
+        neverRan.add(entry.getKey());
+      }
+    }
+    if (activeTasks.isEmpty()) {
+      notifyAll();
+    }
+
+    return neverRan;
+  }
+
+  @Override
+  public synchronized boolean isShutdown() {
+    return shutdown;
+  }
+
+  @Override
+  public synchronized boolean isTerminated() {
+    return shutdown && activeTasks.isEmpty();
+  }
+
+  @Override
+  public synchronized boolean awaitTermination(long timeout, TimeUnit unit)
+      throws InterruptedException {
+    long remainingNanos = unit.toNanos(timeout);
+    long deadline = System.nanoTime() + remainingNanos;
+    while (!isTerminated() && remainingNanos > 0) {
+      TimeUnit.NANOSECONDS.timedWait(this, remainingNanos);
+      remainingNanos = deadline - System.nanoTime();
+    }
+    return isTerminated();
+  }
+
+  @Override
+  public synchronized void execute(final Runnable command) {
+    checkState(!shutdown, "Executor shutting down");
```

Comment:
@carl-mastrangelo it still seems like it might be good to avoid locking around `delegate.submit` if at all possible.  Could we do something like this:

```java
// Step 1: check that we're not shutting down and add it to the 
synchronized (activeTasks) {
    checkState(!shutdown, "Executor shutting down");
    CleanupTask cleanupTask = new CleanupTask(command);
    // Assuming activeTasks is just a set.
    activeTasks.put(cleanupTask);
}

try {
    delegate.submit(cleanupTask);
} catch (Throwable t) {
    activeTasks.done();
    Throwables.propagate(t);
}
```

We could have some sneak in between the synchronized block and the call to execute the task, but they wouldn't get lost since they were added to the list first.  We would just wait for them to complete ... doesn't seem like a big deal.

WDYT?